exporters/jaeger/src/thrift_sender.cc

```diff
@@ -0,0 +1,114 @@
+// Copyright 2021, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "thrift_sender.h"
+#include "udp_transport.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace exporter
+{
+namespace jaeger
+{
+
+using namespace jaegertracing;
+
+ThriftSender::ThriftSender(std::unique_ptr<Transport> &&transport)
+    : transport_(std::move(transport)),
+      protocol_factory_(new apache::thrift::protocol::TCompactProtocolFactory()),
+      thrift_buffer_(new apache::thrift::transport::TMemoryBuffer())
+{}
+
+int ThriftSender::Append(std::unique_ptr<Recordable> &&span) noexcept
+{
+  if (span == nullptr)
+  {
+    return 0;
+  }
+
+  std::lock_guard<std::mutex> guard{lock_};
+
+  uint32_t max_span_bytes = transport_->MaxPacketSize() - kEmitBatchOverhead;
+  if (process_.serviceName.empty())
+  {
+    process_.serviceName = span->ServiceName();
+
+    process_bytes_size_ = CalcSizeOfSerializedThrift(process_);
+    max_span_bytes -= process_bytes_size_;
+  }
+
+  thrift::Span &jaeger_span = *span->Span();
+  jaeger_span.__set_tags(span->Tags());
+
+  const uint32_t span_size = CalcSizeOfSerializedThrift(jaeger_span);
+  if (span_size > max_span_bytes)
+  {
+    // TODO, log too large span error.
+    return 0;
+  }
+
+  byte_buffer_size_ += span_size;
+  if (byte_buffer_size_ <= max_span_bytes)
+  {
+    span_buffer_.push_back(jaeger_span);
+    if (byte_buffer_size_ < max_span_bytes)
+    {
+      return 0;
```

Comment:
Does it mean if there are very few spans created, they may keep waiting for a long time in the queue. Should there be some timeout, or we somehow ensure that every Export() results in the export over the network?

Reply:
Good point. The few events could be there till the tracer object is destructed. We probably could create a separate thread to do flush periodically. But I haven't seen other implementations have the timeout mechanism. How about filing a ticket to follow up?

Reply:
@lalitb great question!

@ThomsonTan this is what OpenTelemetry .NET is doing https://github.com/open-telemetry/opentelemetry-dotnet/blob/98055871511d0b56df75cb7443a1ed9a49918014/src/OpenTelemetry.Exporter.Jaeger/JaegerExporter.cs#L84.

This is the semantic we need to achieve https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#exportbatch.

Reply:
> We probably could create a separate thread to do flush periodically.

This doesn't sound correct to me.

Reply:
Thanks for sharing the spec and OpenTelemetry-DotNet link. The spec clearly requires each batch of spans to be serialized and sent to destination. Add a call to `Flush()` at the end of `Export(batch)` for this.

Reply:
Yes, that seems a good approach.